File: ios-app/Doompedia/Data/DeltaApplier.swift
import Foundation

final class DeltaApplier {
    private let store: SQLiteStore
    private let decoder = JSONDecoder()

    init(store: SQLiteStore) {
        self.store = store
    }

    func apply(from fileURL: URL) throws -> Int {
        if fileURL.pathExtension == "zst" {
            throw NSError(
                domain: "Doompedia",
                code: 51,
                userInfo: [NSLocalizedDescriptionKey: "Compressed delta apply requires decoder integration"]
            )
        }

        var upserts: [SeedRow] = []
        var deletes: [Int64] = []
        var applied = 0

        let lineReader: ((URL, (String) throws -> Void) throws -> Void) = fileURL.pathExtension == "gz"
            ? GzipLineReader.forEachLine
            : LineReader.forEachLine

        try lineReader(fileURL) { line in
            let row = try decoder.decode(DeltaRowPayload.self, from: Data(line.utf8))
            switch row.op {
            case "upsert":
                guard let record = row.record else { return }
                upserts.append(
                    SeedRow(
                        page_id: record.page_id,
                        lang: record.lang,
                        title: record.title,
                        summary: record.summary,
                        wiki_url: record.wiki_url,
                        topic_key: record.topic_key,
                        quality_score: record.quality_score,
                        is_disambiguation: record.is_disambiguation,
                        source_rev_id: record.source_rev_id,
                        updated_at: record.updated_at,
                        aliases: row.aliases ?? []
                    )
                )
                applied += 1

            case "delete":
                if let pageID = row.page_id {
                    deletes.append(pageID)
                    applied += 1
                }

            default:
                throw NSError(
                    domain: "Doompedia",
                    code: 52,
                    userInfo: [NSLocalizedDescriptionKey: "Unknown delta op: \(row.op)"]
                )
            }

            if upserts.count + deletes.count >= 1000 {
                try flush(upserts: &upserts, deletes: &deletes)
            }
        }

        try flush(upserts: &upserts, deletes: &deletes)
        return applied
    }

    private func flush(upserts: inout [SeedRow], deletes: inout [Int64]) throws {
        if !upserts.isEmpty {
            try store.upsertSeedRows(upserts)
            upserts.removeAll(keepingCapacity: true)
        }

        if !deletes.isEmpty {
            try store.deleteArticles(pageIDs: deletes)
            deletes.removeAll(keepingCapacity: true)
        }
    }
}

File: android-app/app/src/main/java/com/feelbachelor/doompedia/data/importer/DeltaApplier.kt
package com.feelbachelor.doompedia.data.importer

import androidx.room.withTransaction
import com.feelbachelor.doompedia.data.db.AliasEntity
import com.feelbachelor.doompedia.data.db.ArticleEntity
import com.feelbachelor.doompedia.data.db.WikiDatabase
import com.feelbachelor.doompedia.domain.normalizeSearch
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.io.File
import java.util.zip.GZIPInputStream

class DeltaApplier(
    private val db: WikiDatabase,
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    suspend fun apply(deltaFile: File): Int {
        return withContext(Dispatchers.IO) {
            val stream = when {
                deltaFile.name.endsWith(".gz") -> GZIPInputStream(deltaFile.inputStream())
                deltaFile.name.endsWith(".zst") -> error("zstd deltas require runtime decoder integration")
                else -> deltaFile.inputStream()
            }

            val upserts = mutableListOf<ArticleEntity>()
            val aliases = mutableListOf<AliasEntity>()
            val updatedPageIds = mutableListOf<Long>()
            val deletes = mutableListOf<Long>()
            var applied = 0

            stream.bufferedReader().useLines { lines ->
                lines.forEach { line ->
                    if (line.isBlank()) return@forEach
                    val row = json.decodeFromString(DeltaRow.serializer(), line)
                    when (row.op) {
                        "upsert" -> {
                            val record = requireNotNull(row.record)
                            upserts += ArticleEntity(
                                pageId = record.page_id,
                                lang = record.lang,
                                title = record.title,
                                normalizedTitle = record.normalized_title.ifBlank { normalizeSearch(record.title) },
                                summary = record.summary,
                                wikiUrl = record.wiki_url,
                                topicKey = TopicClassifier.normalizeTopic(
                                    rawTopic = record.topic_key,
                                    title = record.title,
                                    summary = record.summary,
                                ),
                                qualityScore = record.quality_score,
                                isDisambiguation = record.is_disambiguation,
                                sourceRevId = record.source_rev_id,
                                updatedAt = record.updated_at,
                            )
                            updatedPageIds += record.page_id
                            row.aliases.forEach { alias ->
                                aliases += AliasEntity(
                                    pageId = record.page_id,
                                    lang = record.lang,
                                    alias = alias,
                                    normalizedAlias = normalizeSearch(alias),
                                )
                            }
                            applied++
                        }

                        "delete" -> {
                            val pageId = requireNotNull(row.page_id)
                            deletes += pageId
                            applied++
                        }

                        else -> error("Unknown delta op '${row.op}'")
                    }

                    if (upserts.size + deletes.size >= 1_000) {
                        flush(upserts, aliases, updatedPageIds, deletes)
                    }
                }
            }

            flush(upserts, aliases, updatedPageIds, deletes)
            applied
        }
    }

    private suspend fun flush(
        upserts: MutableList<ArticleEntity>,
        aliases: MutableList<AliasEntity>,
        updatedPageIds: MutableList<Long>,
        deletes: MutableList<Long>,
    ) {
        if (upserts.isEmpty() && deletes.isEmpty()) return

        db.withTransaction {
            val dao = db.wikiDao()
            if (upserts.isNotEmpty()) {
                dao.upsertArticles(upserts)
                dao.deleteAliasesForPages(updatedPageIds)
                if (aliases.isNotEmpty()) {
                    dao.insertAliases(aliases)
                }
            }

            if (deletes.isNotEmpty()) {
                dao.deleteArticles(deletes)
            }
        }

        upserts.clear()
        aliases.clear()
        updatedPageIds.clear()
        deletes.clear()
    }
}
